### PR TITLE
[Error Handling 3] Core Logic

### PR DESCRIPTION
**Pull request by W3stside**
Waterfalls in #746 

Handles the core logic changes of error codes for quote fetch

1. `useRefetchPriceCallback` now deals with the `fee` and `price` promises separately via `Promise.allSettled` 
2. `useRefetchPriceCallback` also handles any errors arising from point 1 and saves to state however necessary based on error codes added in #746
3. removes some legacy code from state and fixes types as necessary

### Testing
To test this, you'll need to use redux inspector in chrome tools to check state

#### Checking `FeeExceedsPrice` error
1. [Mainnet] Select WETH sell (or any idc) against DAI buy
2. Input a tiny amt e.g 0.00000003
3. redux inspector should show, inside state > price: `error: FeeExceedsPrice`
4. app may or may not show that in the swap button state (changes to this are incoming but current code "might just work")

#### Checking `NotFound` aka `InsufficientLiquidity` error
This one is easiest on Rinkeby
1. [Rinkeby] Select WETH sell against SUSD or GUSD buy
2. input any amount in the input fields
3. redux inspector should show, inside state > price: `error: NotFound`
4. app may or may not show "Insufficient liquidity" in the swap button state (changes to this are incoming but current code "might just work")

### CODE REVIEW

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 1]
```diff
@@ -56,18 +53,19 @@ async function _getQuote({ quoteParams, fetchFee, previousFee }: RefetchQuoteCal
...
-          feeExceedsPrice
-        }))
+      ? getPriceQuote({ chainId, baseToken, quoteToken, amount: exchangeAmount, kind })
       : // fee exceeds our price, is invalid
-        Promise.resolve({
-          feeExceedsPrice,
-          token: sellToken,
-          amount: null
-        })
+        Promise.reject(
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:58]: here we reject with the `FeeExceedsPrice` error caught on `L55`

File: src/custom/state/price/reducer.ts [hunk 2]
```diff
@@ -28,8 +28,6 @@ export interface QuoteInformationObject extends FeeQuoteParams {
   price?: PriceInformation
   error?: ApiErrorCodes
   lastCheck: number
-  // TODO: remove
-  feeExceedsPrice: boolean
```
Comment (W3stside) [inline src/custom/state/price/reducer.ts:32]: @anxolin removed here

**Review by alfetopito — APPROVED** [review]

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 3]
```diff
@@ -56,18 +53,19 @@ async function _getQuote({ quoteParams, fetchFee, previousFee }: RefetchQuoteCal
...
+          new OperatorError({
+            errorType: ApiErrorCodes.FeeExceedsFrom,
+            description: OperatorError.apiErrorDetails.FeeExceedsFrom
+          })
+        )
 
-  return Promise.all([pricePromise, feePromise])
+  // Promise.allSettled does NOT throw on 1 promise rejection
```
Comment (anxolin) [inline src/custom/hooks/useRefetchPriceCallback.ts:65]: I think we don't need to explain `Promise.allSettled`

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 4]
```diff
@@ -77,31 +75,51 @@ function _isValidOperatorError(error: any): error is OperatorError {
...
+      }
+      // Fee/Price query returns error
+      // e.g Insufficient Liquidity or Fee exceeds Price
+      case ApiErrorCodes.FeeExceedsFrom:
+      case ApiErrorCodes.NotFound: {
+        console.error(`${error.message}: ${error.description}!`)
+        return setQuoteError({
+          ...quoteData,
+          lastCheck: Date.now(),
+          error: error.type
```
Comment (anxolin) [inline src/custom/hooks/useRefetchPriceCallback.ts:114]: This connect with one comment i did in the PR where I saw you were using here the same error codes as the backend

I think that we should use our own enum, with just the types we handle. Or a string if we can get by with that.
This way, we make sure we translate from backend errors to frontend errors.

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 5]
```diff
@@ -77,31 +75,51 @@ function _isValidOperatorError(error: any): error is OperatorError {
...
+        })
+      }
+      default: {
+        // some other operator error occurred, log it
+        console.error(error)
+        // Clear the quote
+        return clearQuote({ chainId: quoteData.chainId, token: quoteData.sellToken })
```
Comment (anxolin) [inline src/custom/hooks/useRefetchPriceCallback.ts:121]: Shouldn't we at least show that something is wrong? Shouldn't we user `setQuoteError` to inform there was an issue?

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 6]
```diff
@@ -133,8 +155,29 @@ export function useRefetchQuoteCallback() {
...
+        // to save into state and use
+        quoteData = {
+          ...params.quoteParams,
+          fee: getPromiseFulfilledValue(fee, undefined),
+          price: getPromiseFulfilledValue(price, undefined),
+          lastCheck: Date.now()
+        }
+
+        // check the PromiseSettledValue/PromiseRejectedResult
+        // returned by Promise.allSettled
```
Comment (anxolin) [inline src/custom/hooks/useRefetchPriceCallback.ts:173]: Can we try to use a less technical and more descriptive language for the comment?

File: src/custom/state/swap/hooks.ts [hunk 7]
```diff
@@ -25,6 +25,7 @@ import { ParsedQs } from 'qs'
 import { DEFAULT_NETWORK_FOR_LISTS } from 'constants/lists'
 import { WETH_LOGO_URI, XDAI_LOGO_URI } from 'constants/index'
 import { WrappedTokenInfo } from '../lists/hooks'
+import { isFeeGreaterThanPriceError } from '../price/utils'
```
Comment (anxolin) [inline src/custom/state/swap/hooks.ts:28]: absolute path?

File: src/custom/utils/misc.ts [hunk 8]
```diff
@@ -6,6 +6,20 @@ export const isTruthy = <T>(value: T | null | undefined | false): value is T =>
 export const delay = <T = void>(ms = 100, result?: T): Promise<T> =>
   new Promise(resolve => setTimeout(resolve, ms, result))
 
+export function isPromiseFulfilled<T>(
+  promiseResult: PromiseSettledResult<T>
+): promiseResult is PromiseFulfilledResult<T> {
+  return promiseResult.status === 'fulfilled'
+}
+
+// To properly handle PromiseSettleResult which returns and object
```
Comment (anxolin) [inline src/custom/utils/misc.ts:15]: I feel the comment should say what if does, not why you did it. 

sth like `Returns either a fullfilled value or a default`

actually, even the name can hint the responsability `getPrimiseFullfilledOrDefault`

File: src/custom/hooks/useRefetchPriceCallback.ts (hunk 5 above)
Comment (W3stside) [reply src/custom/hooks/useRefetchPriceCallback.ts:121]: likely for another PR but yes

File: src/custom/hooks/useRefetchPriceCallback.ts (hunk 4 above)
Comment (W3stside) [reply src/custom/hooks/useRefetchPriceCallback.ts:114]: ok then i will do that then

Comment (W3stside) [reply src/custom/hooks/useRefetchPriceCallback.ts:114]: notes in https://github.com/gnosis/cowswap/issues/751

File: src/custom/hooks/useRefetchPriceCallback.ts (hunk 6 above)
Comment (W3stside) [reply src/custom/hooks/useRefetchPriceCallback.ts:173]: noted in https://github.com/gnosis/cowswap/issues/751

File: src/custom/state/swap/hooks.ts (hunk 7 above)
Comment (W3stside) [reply src/custom/state/swap/hooks.ts:28]: noted in https://github.com/gnosis/cowswap/issues/751

File: src/custom/utils/misc.ts (hunk 8 above)
Comment (W3stside) [reply src/custom/utils/misc.ts:15]: noted in https://github.com/gnosis/cowswap/issues/751

File: src/custom/hooks/useRefetchPriceCallback.ts (hunk 5 above)
Comment (W3stside) [reply src/custom/hooks/useRefetchPriceCallback.ts:121]: https://github.com/gnosis/cowswap/pull/747#discussion_r647597184

**Pull request merged**